src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -275,53 +286,109 @@ protected override void ProcessRecord()
         protected override void StopProcessing()
         {
             _sender?.SendAsyncCancel();
+            cancellationTokenSource.Cancel();
         }
 
         #region ConnectionTest
 
+        private void SetCountForTcpTest()
+        {
+            if (Repeat.IsPresent)
+            {
+                Count = int.MaxValue;
+            }
+            else if (!MyInvocation.BoundParameters.ContainsKey("Count"))
+            {
+                Count = 1;
+            }
+        }
+
         private void ProcessConnectionByTCPPort(string targetNameOrAddress)
         {
+            SetCountForTcpTest();
+
             if (!TryResolveNameOrAddress(targetNameOrAddress, out _, out IPAddress? targetAddress))
             {
                 return;
             }
 
-            TcpClient client = new TcpClient();
-
-            try
+            for (var i = 1; i <= Count; i++)
             {
-                Task connectionTask = client.ConnectAsync(targetAddress, TcpPort);
-                string targetString = targetAddress.ToString();
+                long latency = 0;
+                SocketError status = SocketError.SocketError;
+
+                Stopwatch stopwatch = new Stopwatch();
 
-                for (var i = 1; i <= TimeoutSeconds; i++)
+                using TcpClient client = new TcpClient();
+
+                try
                 {
-                    Task timeoutTask = Task.Delay(millisecondsDelay: 1000);
-                    Task.WhenAny(connectionTask, timeoutTask).Result.Wait();
+                    stopwatch.Start();
 
-                    if (timeoutTask.Status == TaskStatus.Faulted || timeoutTask.Status == TaskStatus.Canceled)
+                    if (client.ConnectAsync(targetAddress, TcpPort).Wait(TimeoutSeconds * 1000, cancellationTokenSource.Token))
                     {
-                        // Waiting is interrupted by Ctrl-C.
-                        WriteObject(false);
-                        return;
+                        latency = stopwatch.ElapsedMilliseconds;
+                        status = SocketError.Success;
+                    }
+                    else
+                    {
+                        status = SocketError.TimedOut;
                     }
+                }
+                catch (AggregateException ae)
+                {
+                    ae.Handle((ex) =>
+                    {
+                        if (ex is TaskCanceledException)
+                        {
+                            throw new PipelineStoppedException();
+                        }
+                        if (ex is SocketException)
+                        {
+                            SocketException? socketException = ex as SocketException;
+                            status = socketException!.SocketErrorCode;
+                            return true;
+                        }
```

Comment:
```c#
                      if (ex is SocketException socketException)
                      {
                          status = socketException.SocketErrorCode;
                          return true;
                      }
```

Reply:
Ah this is way more elegant! 

I wasn't aware the `is` operator would cast after doing the comparison. Updated!